### ISSUE TITLE
Support integration tests

### ISSUE THREAD

**Issue by slifty** [PhilanthropyDataCommons/service#43]
We're about to need to run integration tests (I imagine many of our tests will rely on DB access in fact!)

This will mean three test commands:

`npm run test:unit` -- runs unit tests (no database)
`npm run test:integration` -- runs untegration tests
`npm run test` -- runs both sets of tests

Part of this issue will be figuring out the right way to do integration tests so that they're run in parallel but don't step on each other.

Potential resources I've found:

https://dev.to/walrusai/testing-database-interactions-with-jest-519n
https://jestjs.io/docs/setup-teardown
https://jestjs.io/docs/configuration#testmatch-arraystring

**Comment by bickelj** [comment]
Is it possible to specify different postgres schemas to create and use when testing rather than the default `public` schema? I don't think any of our migrations or code will refer to a schema. By schema here I mean the postgres construct that lies between `database` and `table`.

**Comment by slifty** [comment]
@bickelj not sure what you mean exactly.

it's possible to mock databases, which I think would be useful in unit tests, but I think integration tests would be more robust using the true pg database + migrations.

(Is any of that useful / responsive to your question?)

**Comment by bickelj** [comment]
@slifty I see that some of the above resources mention creating separate databases for test isolation, which makes sense, like `create database blah owner blah` but there is also the possibility of re-using the same database with separate schemas, such as connect to database blah, but then `create schema test2`, etc. I don't recall if `create schema` is lighter weight or about the same. If it's about the same as `create database` then my suggestion is moot.

**Comment by bickelj** [comment]
@slifty Perhaps one advantage of separate schemas for isolation would be re-use of the connection pool by all tests. Then again, that's a kind of test non-isolation, so it could be a drawback. It would probably be cleanest to have a separate single connection to a separate single database for each test.

**Comment by slifty** [comment]
Ahhhhh got it -- that would be truly excellent.

I'll check out.  At the very least using a separate schema within a database is not having to give "create db" permissions to the test user.

ty for raising it!

**Comment by bickelj** [comment]
I have been trying to get the tests to fail properly when there is a connection failure to the db. On main, I get a timeout failure followed by "there are asynchronous operations that weren't stopped in your tests." I have tried a couple of things but it looks like `beforeEach` is looking for synchronous code, etc. My noobish summary is that I don't think the errors are coming back up the call stack, they're just out there somewhere until some timeouts hit.

**Comment by bickelj** [comment]
See https://github.com/PhilanthropyDataCommons/service/commit/5bdd199af7bae964c14dd980784f7462b2ae8919#r78988507

**Comment by bickelj** [comment]
I have some `beforeEach` and `afterEach` code but it seems like they ignore any errors and continue to execute the test (e.g. timeout instead of exiting and failing).

**Comment by slifty** [comment]
Hmmm interesting -- this does sound pesky.

I wonder if there's a way to (A) signal that beforeEach is async and (B) fail a test from within beforeEach if the db connection is not formed.

I don't think we can avoid the timeout issue (though we could make it easier to configure timeout so that tests can configure a shorter timeout)

meta side note -- what do you think about opening a new issue just for this specific problem (I think we technically have implemented this issue as written).

**Comment by bickelj** [comment]
I have peppered my local modified version of some code associated with PR #31 with logger.debug statements. I think you summarized it well as (A) having to do with `async` functions and (B) having to do with errors during `beforeEach`, but also it looks like something to do with the behavior of (C) TinyPg/Pg.Pool/the-way-we-manage-the-pool/clients. I will create an issue more precisely describing the issue.

**Comment by bickelj** [comment]
Created #53 with some snippets.

**Comment by slifty** [comment]
Thank you!  I have one small patch I'm going to make to this issue (the bug you found notwithstanding) -- which is to have the harness functions called by the suite (apparently Jest stops test flow if there is a failure, and so `afterEach` and `beforeEach` really do need to be specified to ensure everything is set up and cleaned up regardless of test outcome.
